fix(dataset): Respect columns configuration when duplicating datasets

Antonio-RiveroMartnez opened the pull request:
### SUMMARY

When duplicating datasets, we currently are forcing columns on the new dataset to have `groupby` and `filreable` set to `True` even if the user marked those as false in the base dataset. With this changes we are fixing that. Also did a small refactor in how we raise exceptions so they are all in the `validate` method and using existing exceptions.


### BEFORE/AFTER SCREENSHOTS OR ANIMATED GIF
Before:
![error](https://user-images.githubusercontent.com/38889534/221206730-04342723-704e-4c88-a8e4-4d9d1249551e.gif)


After:
![test](https://user-images.githubusercontent.com/38889534/221206798-6242c12c-5bca-46d7-a718-22000edae1fe.gif)


### TESTING INSTRUCTIONS
1. Create a Virtual dataset
2. Create a `calculated column` in your new dataset
3. Check `isFilterable` for the new column and save the changes
4. Open the new dataset for edition and uncheck `isFilterable` so it is False
5. Duplicate the dataset
6. The new dataset must have `isFilterable` unchecked

### ADDITIONAL INFORMATION
<!--- Check any relevant boxes with "x" -->
<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->
- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

Antonio-RiveroMartnez committed:
Dataset Duplication:

- Keep validation exceptions in the validate method
- Stop setting filterable and groupby to True for duplicated columns, they must respect base configuration

Antonio-RiveroMartnez commented on superset/datasets/commands/duplicate.py:
```diff
@@ -44,60 +43,51 @@
 class DuplicateDatasetCommand(CreateMixin, BaseCommand):
     def __init__(self, data: Dict[str, Any]) -> None:
-        self._base_model: SqlaTable = SqlaTable()
+        self._base_model: Optional[SqlaTable] = None
```
Changed to `Optional` since `find_by_id` could return `None`, and also it's more consistent how we handle such cases.

Antonio-RiveroMartnez commented on superset/datasets/commands/duplicate.py:
```diff
@@ -44,60 +43,51 @@
-                mets.append(met)
-            table.metrics = mets
-            db.session.commit()
+            if self._base_model and self._database:
```
Forced to use a check before so `pre-commit` doesn't complain when assigning the properties to the new `dataset`. The majority of the code inside this block remains untouched.

eschutho replied:
@betodealmeida or @hughhhh what do you think here? The validate method will raise if self._base_model or self._database are falsy. Technically those values couldn't be false on line 53. I'm leaning for readability sake to just skip the pypy warning so that the reader understands that those values couldn't be falsy here.

Antonio-RiveroMartnez replied:
Hey @eschutho I changed it now, based on your comments. Thanks

Antonio-RiveroMartnez commented on superset/datasets/commands/duplicate.py:
```diff
@@ -44,60 +43,51 @@
-            owners = self._properties["owners"]
-            database = db.session.query(Database).get(database_id)
-            if not database:
-                raise SupersetErrorException(
```
Changed to use `DatabaseNotFoundError` instead of constructing this here.

Antonio-RiveroMartnez committed:
Dataset Duplication:

- Remove conditional at the top and instead ignore type validation for certain line to improve code readability

Antonio-RiveroMartnez commented on superset/datasets/commands/duplicate.py:
```diff
@@ -44,49 +43,39 @@
-            table.database = database
-            table.schema = self._base_model.schema
-            table.template_params = self._base_model.template_params
+            table.database = self._database  # type: ignore
```
Addressing review comments, we are now leaning for readability.

eschutho replied:
@Antonio-RiveroMartnez thanks for this! I've asked @betodealmeida to see if he can suggest any other patterns here, since now based on my feedback there are a lot of `ignore`s. We talked about validating sooner, or moving some of the validation logic into the init. If there are any suggestions that you have to make this cleaner, we can propose a new pattern here.

Antonio-RiveroMartnez committed:
Dataset Duplication:

- Use assert to let users know that database and dataset are required to continue further, this also prevent the usage of type ignores
